Handle `export {App}`, not only `export default App` from exercises

Workshop exercises (and final) have to use default exports to support `React.lazy` (which can only render a default export) in https://github.com/kentcdodds/react-workshop-app/blob/7a5504500fc75c7275933ca1e16bc30fab45bd6e/src/index.tsx#L66

However, it can be confusing if someone tries to import a function that is not exported, e.g. `import Greeting from ...` while the exported function is `export default App` - see https://discord.com/channels/715220730605731931/759594515668926514/802835578398965780

I noticed that Kent prefers named exports anyway and it should be possible hammer the Promise with `import().then( ??? )` into a shape supported by React.lazy, so that the exercises can be rewritten into `export {App}` => I will try to make PR and we can discuss whether such hack is useful enough..

I'm thinking to support following in exercise files:
1. `export {App}` and/or `export function App` as the new standard
2. `export default Whatever` for gradual transition... (maybe with a `console.warn` in Jest, as a gentle reminder between TypeScriptifying and re-recording videos)
3. if there is no default export and no App, render an error message that a component called `App` has to be exported from the exercise file

I was able to make the unit tests green, but `npm run dev`  does not work on Windows for me - probably due to unfinished migration of `codegen.macro` to typesript (the generated JS file contains `require('./load-files')` which works fine when executed directly in Node in the `dist` directory, but seems to be problematic via codegen for unknown reason).

```js
./src/index.js
Error: C:\Users\Peter\git\react-workshop-app\example\react-fundamentals\src\index.js: codegen.macro:
 Cannot find module './load-files'

Require stack:
- C:\Users\Peter\git\react-workshop-app\dist\get-app-info.js
- C:\Users\Peter\git\react-workshop-app\dist\codegen.js
- C:\Users\Peter\git\react-workshop-app\codegen.js
- C:\Users\Peter\git\react-workshop-app\example\react-fundamentals\src\index.js
- C:\Users\Peter\git\react-workshop-app\example\react-fundamentals\node_modules\babel-plugin-codegen\dist\helpers.js
- C:\Users\Peter\git\react-workshop-app\example\react-fundamentals\node_modules\babel-plugin-codegen\dist\replace.js
- C:\Users\Peter\git\react-workshop-app\example\react-fundamentals\node_modules\babel-plugin-codegen\dist\macro.js
- C:\Users\Peter\git\react-workshop-app\example\react-fundamentals\node_modules\babel-plugin-codegen\macro.js
- C:\Users\Peter\git\react-workshop-app\example\react-fundamentals\node_modules\codegen.macro\index.js
- C:\Users\Peter\git\react-workshop-app\example\react-fundamentals\node_modules\babel-preset-react-app\node_modules\babel-plugin-macros\dist\index.js
- C:\Users\Peter\git\react-workshop-app\example\react-fundamentals\node_modules\babel-preset-react-app\create.js
- C:\Users\Peter\git\react-workshop-app\example\react-fundamentals\node_modules\babel-preset-react-app\index.js
- C:\Users\Peter\git\react-workshop-app\example\react-fundamentals\node_modules\@babel\core\lib\config\files\plugins.js
- C:\Users\Peter\git\react-workshop-app\example\react-fundamentals\node_modules\@babel\core\lib\config\files\index.js
- C:\Users\Peter\git\react-workshop-app\example\react-fundamentals\node_modules\@babel\core\lib\index.js
- C:\Users\Peter\git\react-workshop-app\example\react-fundamentals\node_modules\rollup-plugin-babel\dist\rollup-plugin-babel.cjs.js
- C:\Users\Peter\git\react-workshop-app\example\react-fundamentals\node_modules\workbox-build\build\lib\bundle.js
- C:\Users\Peter\git\react-workshop-app\example\react-fundamentals\node_modules\workbox-webpack-plugin\build\generate-sw.js
- C:\Users\Peter\git\react-workshop-app\example\react-fundamentals\node_modules\workbox-webpack-plugin\build\index.js
- C:\Users\Peter\git\react-workshop-app\example\react-fundamentals\node_modules\react-scripts\config\webpack.config.js
- C:\Users\Peter\git\react-workshop-app\example\react-fundamentals\node_modules\react-scripts\scripts\start.js
Learn more: https://www.npmjs.com/package/codegen.macro
```

--- a/.gitignore
+++ b/.gitignore
@@ -15,3 +15,4 @@ cypress/videos
 # when working with contributors
 package-lock.json
 yarn.lock
+.eslintcache

--- a/example/react-fundamentals/src/__tests__/05.js
+++ b/example/react-fundamentals/src/__tests__/05.js
@@ -1,15 +1,15 @@
 import React from 'react'
 import chalk from 'chalk'
 import {render, prettyDOM} from '@testing-library/react'
-import Usage from '../final/05'
-// import Usage from '../exercise/05'
+import {App} from '../final/05'
+// import App from '../exercise/05'
 
 test('renders the correct styles new', () => {
-  const {container, getByText, getAllByText} = render(<Usage />)
+  const {container, getByText, getAllByText} = render(<App />)
   const allBoxes = getAllByText(/box/i)
 
   try {
-    allBoxes.forEach(box => expect(box).toHaveClass('box'))
+    allBoxes.forEach((box) => expect(box).toHaveClass('box'))
   } catch (error) {
     //
     //
@@ -24,7 +24,7 @@ test('renders the correct styles new', () => {
   }
 
   try {
-    allBoxes.forEach(box => expect(box).toHaveStyle('font-style: italic;'))
+    allBoxes.forEach((box) => expect(box).toHaveStyle('font-style: italic;'))
   } catch (error) {
     //
     //

--- a/example/react-fundamentals/src/exercise/05.tsx
+++ b/example/react-fundamentals/src/exercise/05.tsx
@@ -43,4 +43,10 @@ function Usage() {
   )
 }
 
-export default Usage
+// testing a "forgotten" export
+// export default Usage
+
+/*
+eslint
+  @typescript-eslint/no-unused-vars: "off"
+*/

--- a/example/react-fundamentals/src/final/05.tsx
+++ b/example/react-fundamentals/src/final/05.tsx
@@ -28,7 +28,7 @@ const largeBox = (
   </div>
 )
 
-function Usage() {
+function App() {
   return (
     <div>
       {smallBox}
@@ -38,4 +38,4 @@ function Usage() {
   )
 }
 
-export default Usage
+export {App}

--- a/src/__tests__/index.js
+++ b/src/__tests__/index.js
@@ -7,6 +7,8 @@ import {screen, fireEvent} from '@testing-library/react'
 import {makeKCDWorkshopApp} from '..'
 import {getAppInfo} from '../get-app-info'
 
+const consoleErrorMock = jest.spyOn(console, 'error').mockImplementation()
+
 beforeAll(() => {
   matchMediaPolyfill(window)
 })
@@ -24,6 +26,7 @@ afterEach(() => {
   // so we'll reset the innerHTML
   // document.body.removeChild(root)
   document.body.innerHTML = ''
+  consoleErrorMock.mockClear()
 })
 
 function setup() {
@@ -32,10 +35,19 @@ function setup() {
   })
 
   const imports = {
-    'src/exercise/05.tsx': () =>
-      import('../../example/react-fundamentals/src/exercise/05.tsx'),
+    'src/exercise/01.md': () =>
+      import('../../example/react-fundamentals/src/exercise/01.md'),
+    'src/exercise/01.html': () =>
+      import('../../example/react-fundamentals/src/exercise/01.html'),
+    'src/final/01.html': () =>
+      import('../../example/react-fundamentals/src/final/01.html'),
+    'src/final/01.extra-1.html': () =>
+      import('../../example/react-fundamentals/src/final/01.extra-1.html'),
+
     'src/exercise/05.md': () =>
       import('../../example/react-fundamentals/src/exercise/05.md'),
+    'src/exercise/05.tsx': () =>
+      import('../../example/react-fundamentals/src/exercise/05.tsx'),
     'src/final/05.tsx': () =>
       import('../../example/react-fundamentals/src/final/05.tsx'),
     'src/final/05.extra-1.tsx': () =>
@@ -66,6 +78,8 @@ test('regular app', () => {
       name: /styling/i,
     }),
   )
+
+  expect(consoleErrorMock).toHaveBeenCalledTimes(0)
 })
 
 test('isolated page', async () => {
@@ -82,4 +96,25 @@ test('isolated page', async () => {
   })
 
   await screen.findByText('large orange box')
+
+  // isolated page seems to execute React.lazy of unrelated imports => no console error expected if fixed
+  expect(consoleErrorMock).toHaveBeenCalledTimes(1)
+})
+
+test('isolated page with forgotten export', async () => {
+  const {imports, filesInfo, gitHubRepoUrl} = setup()
+
+  window.history.pushState({}, 'Test page', '/isolated/exercise/05.tsx')
+
+  makeKCDWorkshopApp({
+    imports,
+    filesInfo,
+    gitHubRepoUrl,
+    projectTitle: 'test project',
+    backend: require('../../example/react-fundamentals/src/backend'),
+  })
+
+  await screen.findByText(/add `export {App}`/)
+
+  expect(consoleErrorMock).toHaveBeenCalledTimes(2)
 })

--- a/src/__tests__/index.js
+++ b/src/__tests__/index.js
@@ -26,22 +26,35 @@ afterEach(() => {
   document.body.innerHTML = ''
 })
 
-test('regular app', () => {
-  const {filesInfo, gitHubRepoUrl} = getAppInfo({
+function setup() {
+  const {filesInfo: filesInfoRaw, gitHubRepoUrl} = getAppInfo({
     cwd: path.join(process.cwd(), 'example/react-fundamentals'),
   })
 
+  const imports = {
+    'src/exercise/05.tsx': () =>
+      import('../../example/react-fundamentals/src/exercise/05.tsx'),
+    'src/exercise/05.md': () =>
+      import('../../example/react-fundamentals/src/exercise/05.md'),
+    'src/final/05.tsx': () =>
+      import('../../example/react-fundamentals/src/final/05.tsx'),
+    'src/final/05.extra-1.tsx': () =>
+      import('../../example/react-fundamentals/src/final/05.extra-1.tsx'),
+  }
+  const filesInfo = filesInfoRaw.filter(({id}) => imports[id])
+
+  expect(filesInfo.map(({id}) => id).sort()).toEqual(
+    Object.keys(imports).sort(),
+  )
+
+  return {imports, filesInfo, gitHubRepoUrl}
+}
+
+test('regular app', () => {
+  const {imports, filesInfo, gitHubRepoUrl} = setup()
+
   makeKCDWorkshopApp({
-    imports: {
-      'src/exercise/05.js': () =>
-        import('../../example/react-fundamentals/src/exercise/05.js'),
-      'src/exercise/05.md': () =>
-        import('../../example/react-fundamentals/src/exercise/05.md'),
-      'src/final/05.js': () =>
-        import('../../example/react-fundamentals/src/final/05.js'),
-      'src/final/05.extra-1.js': () =>
-        import('../../example/react-fundamentals/src/final/05.extra-1.js'),
-    },
+    imports,
     filesInfo,
     gitHubRepoUrl,
     projectTitle: 'test project',
@@ -55,25 +68,13 @@ test('regular app', () => {
   )
 })
 
-// TODO: fix this test
-test.skip('isolated page', async () => {
-  const {filesInfo, gitHubRepoUrl} = getAppInfo({
-    cwd: path.join(process.cwd(), 'example/react-fundamentals'),
-  })
+test('isolated page', async () => {
+  const {imports, filesInfo, gitHubRepoUrl} = setup()
 
-  window.history.pushState({}, 'Test page', '/isolated/final/05.js')
+  window.history.pushState({}, 'Test page', '/isolated/final/05.tsx')
 
   makeKCDWorkshopApp({
-    imports: {
-      'src/exercise/05.js': () =>
-        import('../../example/react-fundamentals/src/exercise/05.js'),
-      'src/exercise/05.md': () =>
-        import('../../example/react-fundamentals/src/exercise/05.md'),
-      'src/final/05.js': () =>
-        import('../../example/react-fundamentals/src/final/05.js'),
-      'src/final/05.extra-1.js': () =>
-        import('../../example/react-fundamentals/src/final/05.extra-1.js'),
-    },
+    imports,
     filesInfo,
     gitHubRepoUrl,
     projectTitle: 'test project',
@@ -82,8 +83,3 @@ test.skip('isolated page', async () => {
 
   await screen.findByText('large orange box')
 })
-
-/*
-eslint
-  import/no-unresolved: "off",
-*/

--- a/src/codegen.tsx
+++ b/src/codegen.tsx
@@ -3,17 +3,12 @@ import {getAppInfo} from './get-app-info'
 
 function getCode({
   cwd = process.cwd(),
-  ignore,
   options,
 }: {
   cwd?: string
-  ignore?: Array<string>
   options?: Record<string, unknown>
 } = {}) {
-  const {gitHubRepoUrl, filesInfo, hasBackend, imports} = getAppInfo({
-    cwd,
-    ignore,
-  })
+  const {gitHubRepoUrl, filesInfo, hasBackend, imports} = getAppInfo({cwd})
   return `
 import {makeKCDWorkshopApp} from '@kentcdodds/react-workshop-app'
 import {loadDevTools} from '@kentcdodds/react-workshop-app/dev-tools'

--- a/src/get-app-info.tsx
+++ b/src/get-app-info.tsx
@@ -4,16 +4,13 @@ import path from 'path'
 import {loadFiles} from './load-files'
 import type {FileInfo} from './types'
 
-function getAppInfo({
-  cwd = process.cwd(),
-  ignore,
-}: {cwd?: string; ignore?: Array<string>} = {}): {
+function getAppInfo({cwd = process.cwd()}: {cwd?: string} = {}): {
   gitHubRepoUrl: string
   filesInfo: Array<FileInfo>
   imports: Array<string>
   hasBackend: boolean
 } {
-  const filesInfo = loadFiles({cwd, ignore})
+  const filesInfo = loadFiles({cwd})
   let gitHubRepoUrl
   const pkgPath = path.join(process.cwd(), 'package.json')
   try {

--- a/src/index.tsx
+++ b/src/index.tsx
@@ -11,6 +11,7 @@ import type {
   Imports,
   Backend,
   DynamicImportFn,
+  DefaultDynamicImportFn,
 } from './types'
 
 const styleTag = document.createElement('style')
@@ -63,7 +64,9 @@ function makeKCDWorkshopApp({
 
   for (const {ext, filePath} of filesInfo) {
     if (componentExtensions.includes(ext)) {
-      lazyComponents[filePath] = React.lazy(imports[filePath])
+      lazyComponents[filePath] = React.lazy(
+        moduleWithDefaultExport(imports, filePath),
+      )
     }
   }
 
@@ -166,7 +169,7 @@ function makeKCDWorkshopApp({
     }, 20)
 
     if (isIsolated && info) {
-      renderIsolated(imports[info.filePath])
+      renderIsolated(moduleWithDefaultExport(imports, info.filePath))
     } else if (previousIsIsolated !== isIsolated) {
       // if we aren't going from isolated to the app, then we don't need
       // to bother rendering react anew. The app will handle that.
@@ -185,7 +188,7 @@ function makeKCDWorkshopApp({
 
     void isolatedModuleImport().then(async ({default: defaultExport}) => {
       if (history.location !== previousLocation) {
-        // locaiton has changed while we were getting the module
+        // location has changed while we were getting the module
         // so don't bother doing anything... Let the next event handler
         // deal with it
         return
@@ -279,11 +282,40 @@ function makeKCDWorkshopApp({
   handleLocationChange()
 }
 
+function moduleWithDefaultExport(imports: Imports, filePath: string) {
+  const importFn = imports[filePath]
+  if (!importFn) throw new Error(`'${filePath}' does not exist in imports.`)
+
+  if (filePath.match(/\.mdx?$/)) return importFn as DefaultDynamicImportFn
+
+  const promiseOfModule = importFn()
+  const promiseOfDefault: ReturnType<DefaultDynamicImportFn> = promiseOfModule
+    .then(module => {
+      const Component = module.App ?? module.default
+      if (!Component) {
+        throw Error(
+          'Please add a `export {App}` or `export default ...` to your exercise file to render it.',
+        )
+      }
+      return {default: Component}
+    })
+    .catch(error => {
+      console.error(filePath, error)
+      return {
+        default: () => {
+          throw error
+        },
+      }
+    })
+
+  return () => promiseOfDefault
+}
 export {makeKCDWorkshopApp}
 
 /*
 eslint
   babel/no-unused-expressions: "off",
   @typescript-eslint/no-explicit-any: "off",
+  @typescript-eslint/prefer-regexp-exec: "off",
   no-void: "off"
 */

--- a/src/index.tsx
+++ b/src/index.tsx
@@ -286,25 +286,24 @@ function moduleWithDefaultExport(imports: Imports, filePath: string) {
   const importFn = imports[filePath]
   if (!importFn) throw new Error(`'${filePath}' does not exist in imports.`)
 
-  if (filePath.match(/\.mdx?$/)) return importFn as DefaultDynamicImportFn
+  if (filePath.match(/\.(mdx?|html)$/))
+    return importFn as DefaultDynamicImportFn
 
   const promiseOfModule = importFn()
   const promiseOfDefault: ReturnType<DefaultDynamicImportFn> = promiseOfModule
     .then(module => {
       const Component = module.App ?? module.default
-      if (!Component) {
-        throw Error(
-          'Please add a `export {App}` or `export default ...` to your exercise file to render it.',
+      if (typeof Component !== 'function') {
+        throw new Error(
+          'Please add `export {App}` or `export default Component` to your exercise file to render it.',
         )
       }
       return {default: Component}
     })
-    .catch(error => {
+    .catch((error: Error) => {
       console.error(filePath, error)
       return {
-        default: () => {
-          throw error
-        },
+        default: () => <div>{error.message}</div>,
       }
     })
 
@@ -317,5 +316,6 @@ eslint
   babel/no-unused-expressions: "off",
   @typescript-eslint/no-explicit-any: "off",
   @typescript-eslint/prefer-regexp-exec: "off",
+  react/jsx-no-useless-fragment: "off",
   no-void: "off"
 */

--- a/src/load-files.tsx
+++ b/src/load-files.tsx
@@ -5,7 +5,7 @@ import type {FileInfo} from './types'
 
 function loadFiles({cwd = process.cwd(), ...rest} = {}): Array<FileInfo> {
   const fileInfo = glob
-    .sync('src/{exercise,final}/*.+(js|html|jsx|ts|tsx|md|mdx)', {cwd, ...rest})
+    .sync('src/{exercise,final,examples}/*.+(js|html|jsx|ts|tsx|md|mdx)', {cwd, ...rest})
     // eslint-disable-next-line complexity
     .map(filePath => {
       const fullFilePath = path.join(cwd, filePath)

--- a/src/load-files.tsx
+++ b/src/load-files.tsx
@@ -3,20 +3,9 @@ import fs from 'fs'
 import glob from 'glob'
 import type {FileInfo} from './types'
 
-function loadFiles({
-  cwd = process.cwd(),
-  ignore = [
-    '**/__tests__/**',
-    '**/test/**',
-    '**/backend.+(js|ts|tsx)',
-    '**/setupTests.+(js|ts|tsx)',
-    '**/setupProxy.+(js|ts|tsx)',
-    '**/*.d.ts',
-  ],
-  ...rest
-} = {}): Array<FileInfo> {
+function loadFiles({cwd = process.cwd(), ...rest} = {}): Array<FileInfo> {
   const fileInfo = glob
-    .sync('src/**/*.+(js|html|jsx|ts|tsx|md|mdx)', {cwd, ignore, ...rest})
+    .sync('src/{exercise,final}/*.*', {cwd, ...rest})
     // eslint-disable-next-line complexity
     .map(filePath => {
       const fullFilePath = path.join(cwd, filePath)

--- a/src/load-files.tsx
+++ b/src/load-files.tsx
@@ -5,7 +5,7 @@ import type {FileInfo} from './types'
 
 function loadFiles({cwd = process.cwd(), ...rest} = {}): Array<FileInfo> {
   const fileInfo = glob
-    .sync('src/{exercise,final}/*.*', {cwd, ...rest})
+    .sync('src/{exercise,final}/*.+(js|html|jsx|ts|tsx|md|mdx)', {cwd, ...rest})
     // eslint-disable-next-line complexity
     .map(filePath => {
       const fullFilePath = path.join(cwd, filePath)

--- a/src/types.ts
+++ b/src/types.ts
@@ -15,11 +15,23 @@ type FileInfo = {
   extraCreditTitle?: string
 }
 
-type LazyComponents = Record<string, React.LazyExoticComponent<any> | undefined>
+type NoPropsComponent = React.ComponentType<{}>
 
-type DynamicImportFn = () => Promise<{default: React.ComponentType<unknown>}>
+type LazyComponents = Record<
+  string,
+  React.LazyExoticComponent<NoPropsComponent> | undefined
+>
 
-type Imports = Record<string, DynamicImportFn>
+type DynamicImportFn = () => Promise<{
+  App?: NoPropsComponent
+  default?: NoPropsComponent
+}>
+
+type DefaultDynamicImportFn = () => Promise<{
+  default: NoPropsComponent
+}>
+
+type Imports = Record<string, DynamicImportFn | undefined>
 
 type Backend = {
   handlers: Array<RequestHandler>
@@ -30,7 +42,15 @@ type Backend = {
   [key: string]: unknown
 }
 
-export type {FileInfo, LazyComponents, Imports, Backend, DynamicImportFn}
+export type {
+  FileInfo,
+  LazyComponents,
+  Imports,
+  Backend,
+  NoPropsComponent,
+  DynamicImportFn,
+  DefaultDynamicImportFn,
+}
 
 /*
 eslint

